### PR TITLE
Add support for VFS addons

### PR DESCRIPTION
**Pull request by notspiff**
This is my VFS add-on work. I send the PR to see if there is interest.

Some remaining stuff;
- get the datetime things into kodiplatform (see commit disabling it for now).
- clean up the callbacks. note; we have to do it like this due to threading stuff - we cannot just use the normal add-on callbacks.

see the repo list at https://github.com/notspiff for some example add-ons. No guarantee many of them are in working condition, i know vfs.sacd is (in terms of the iface), as well as vfs.rar, but there's some stability issues in the latter. also quite a few are now probably void since you have removed the code from mainline.

### CODE REVIEW

File: xbmc/addons/Addon.cpp [hunk 1]
```diff
@@ -412,6 +413,8 @@ void CAddon::BuildLibName(const cp_extension_t *extension)
       case ADDON_SERVICE:
       case ADDON_REPOSITORY:
       case ADDON_CONTEXT_ITEM:
+      case ADDON_AUDIODECODER:
```
Comment (Jalle19) [inline xbmc/addons/Addon.cpp]: Where did this come from?

Comment (notspiff) [inline xbmc/addons/Addon.cpp]: 3 years of constant rebasing, where things originating in my branch has been pushed to mainline in slightly altered forms...

**Comment by FernetMenta** [comment]
IMO an excellent idea. Once this is integrated we can start cleaning up vfs, i.e. PVR is the only client of LiveTV interface and can make use of this extension point.

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp [hunk 2]
```diff
@@ -36,14 +36,25 @@
 #include "storage/MediaManager.h"
 #include "URL.h"
 #include "filesystem/File.h"
+#include "utils/StringUtils.h"
 #include "utils/URIUtils.h"
+#include "addons/VFSEntry.h"
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp]: Those changes here look like they don't belong to this PR.

Comment (notspiff) [inline xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp]: they do, sort of. i have an add-on adding support for additional .iso files (SACD's). to distinguish these from the video iso images, the content type needs to be set on the item probed. if not it falls back to generic decision based on the fact that the extension is '.iso'.

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp]: why do you include AddonManager.h and VFSEntry.h? namespace ADDON is not used either.

Comment (notspiff) [inline xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp]: ah, right. yes the includes should be removed.

File: xbmc/addons/VFSEntry.cpp [hunk 3]
```diff
@@ -0,0 +1,666 @@
...
+    VFSDirEntriesToCFileItemList(num_entries, entries, items);
+    m_pStruct->FreeDirectory(ctx2);
+
+    return true;
+  }
+
+  return false;
+}
+
+bool CVFSEntry::ContainsFiles(const CURL& url, CFileItemList& items)
```
Comment (Jalle19) [inline xbmc/addons/VFSEntry.cpp]: It seems this method does a lot of different things, any chance it could be split up?

Comment (notspiff) [inline xbmc/addons/VFSEntry.cpp]: Not without changing the vfs api.

Comment (Jalle19) [inline xbmc/addons/VFSEntry.cpp]: Okay

File: addons/resource.language.en_gb/resources/strings.po [hunk 4]
```diff
@@ -876,7 +876,9 @@ msgctxt "#203"
 msgid "Plot outline"
 msgstr ""
 
-#empty string with id 204
+msgctxt "#204"
```
Comment (da-anda) [inline addons/resource.language.en_gb/resources/strings.po]: mind adding a contextual note and also the place where it's used? Something like:
```
#. Name of an add-on type/category
#:  xbmc/addons/Addon.cpp
```
Also, we prefer higher string IDs these days (5 digits), because lower numbers will give way to much wrong hits when searching the code for certain IDs. Thanks.

File: xbmc/addons/AddonCallbacksAddon.cpp [hunk 5]
```diff
@@ -543,6 +543,7 @@ static void CFileItemListToVFSDirEntries(VFSDirEntry* entries,
     entries[i].label = strdup(items[i]->GetLabel().c_str());
     entries[i].path = strdup(items[i]->GetPath().c_str());
     entries[i].size = items[i]->m_dwSize;
+    //entries[i].mtime = items[i]->m_dateTime;
```
Comment (garbear) [inline xbmc/addons/AddonCallbacksAddon.cpp]: Rebase error? The time field was removed from [kodi_vfs_types.h](https://github.com/xbmc/xbmc/blob/master/xbmc/addons/include/kodi_vfs_types.h)

File: xbmc/addons/AddonManager.cpp [hunk 6]
```diff
@@ -37,6 +37,10 @@
 #include "DllLibCPluff.h"
 #include "events/AddonManagementEvent.h"
 #include "events/EventLog.h"
+#include "utils/StringUtils.h"
+#include "utils/JobManager.h"
+#include "threads/SingleLock.h"
+#include "VFSEntry.h"
```
Comment (garbear) [inline xbmc/addons/AddonManager.cpp]: Looks like another rebase error. includes not needed

File: xbmc/addons/VFSEntry.cpp [hunk 7]
```diff
@@ -0,0 +1,666 @@
...
+
+  return m_pStruct->GetStartTime(ctx);
+}
+
+bool CVFSEntry::UpdateItem(void* ctx, CFileItem& item)
+{
+  if (!Initialized())
+    return false;
+
+  return false;//m_pStruct->UpdateItem(ctx);
```
Comment (garbear) [inline xbmc/addons/VFSEntry.cpp]: Intended?

File: xbmc/addons/VFSEntry.cpp [hunk 8]
```diff
@@ -0,0 +1,666 @@
...
+}
+
+void CVFSEntryIDirectoryWrapper::SetErrorDialog2(const char* heading,
+                                                 const char* line1,
+                                                 const char* line2,
+                                                 const char* line3)
+{
+  CVariant l2=0, l3=0;
+  if (line2)
+    l2 = std::string(line2);
```
Comment (garbear) [inline xbmc/addons/VFSEntry.cpp]: is line2 null-checked before this line? I traced the execution back to the DLL, so we should sanity-check its values somewhere (probably as soon as it leaves the DLL boundary)

File: xbmc/addons/AddonCallbacksAddon.cpp (hunk 5 above)
Comment (notspiff) [inline xbmc/addons/AddonCallbacksAddon.cpp]: uhm rewriting history? stuff came from me remember. why do you think it was there in the first place?

upstream in all its viceness removed my datetime stuff from kodiplatform. that is the only rebase error involved. it was removed from kodi_vfs_types because it was unused.

**Comment by garbear** [comment]
@notspiff this PR was looking pretty good, what's the status? I'm happy to help if you have any TODOs left

**Comment by notspiff** [comment]
i haven't touched it. there were some comments, feel free to address those. i will try to find some time..

**Comment by notspiff** [comment]
rebased and updated. very little tested, in particular i'm not sure how well BinaryAddonCache manages the lifetime of add-ons.

example addon: https://github.com/notspiff/vfs.sacd

this is for SuperAudioCD isos. that is the reason for the last commit.

the hack commit highlights a nasty issue with current code design, which is triggered by adding this code.

**Comment by garbear** [comment]
thanks for the rebase. it'll probably be a few months before I pull this into my retroplayer branch, i'm focusing on getting input near flawless for the v17 release.

**Comment by notspiff** [comment]
Rather it goes to master. Retro carries enough /random as-is.

19. aug. 2016 17:17 skrev "Garrett Brown" <notifications@github.com>:

**Comment by garbear** [comment]
Rather RetroPlayer goes to master ;)

**Comment by akva2** [comment]
modernized https://github.com/notspiff/vfs.rar

added a commit that removes the internal rar support for this reason. the add-on will not work if core carries UnrarXLib.

**Comment by akva2** [comment]
modernized https://github.com/notspiff/vfs.sftp

added a commit that removes the internal sftp support.

**Comment by bam80** [comment]
What is left to merge it?

**Comment by akva2** [comment]
(notspiff under his day-work intentity); from my POV:

0) not being in a feature freeze
1) merging the awesome work of @AlwinEsch which cleans up the whole binary add-on concept in kodi, see https://github.com/AlwinEsch/kodi/pull/28
2) rewriting the API to use the new constructs and adding error handling and such (largely "automated" by using the new API constructs).
3) choosing a single add-on to focus on, preferably something that isn't new. like vfs.sftp. then remove remove the internal and have things real-world tested for a while. if all goes well, move more shit to add-ons / start adding new stuff like vfs.sacd

**Comment by itouch5000** [comment]
Any news on this topic? Will this change be part of Kodi 18?

**Comment by notspiff** [comment]
rebased.

**Comment by wsnipex** [comment]
jenkins build this with addons please

**Comment by notspiff** [comment]
ish, sorry for bad timing. i was asked to remove the removal commits.

**Comment by Paxxi** [comment]
jenkins build this with addons please

**Comment by Paxxi** [comment]
no relevant failures in jenkins.

File: xbmc/ServiceManager.h [hunk 9]
```diff
@@ -72,6 +72,7 @@ class CServiceManager
   CPlatform& GetPlatform();
 
   PLAYLIST::CPlayListPlayer& GetPlaylistPlayer();
+  int init_level = 0;
```
Comment (ksooo) [inline xbmc/ServiceManager.h]: Is there a reason for this being a public member?

Why is the member decl placed randomly in between member function's decl instead of having it where the other members are? 

And why has it such a "strange" name? Not 100% sure, but don't we usually use the same notation for members, would be <code>iInitLevel</code> in this case?

Comment (ksooo) [inline xbmc/ServiceManager.h]: > this is not meant for inclusion as-is just highlights the issue.

 hmm.

Comment (notspiff) [inline xbmc/ServiceManager.h]: as i tried to explain in the commit message and the original PR description, this was meant to highlight the issue (using reference returns for things that is dynamically allocated). i just did what was necessary to make it work again and variable names was not a focus point ;)

Comment (notspiff) [inline xbmc/ServiceManager.h]: oh, and the issue already exists in mainline, it just doesn't bite that often (you need to have a file type handled by a audio decoder with multitrack support sitting in some directory listed at early stages in the application). it bit me with rar cause i had some archive in a temp dir which made me notice..
